src/snowflake/cli/plugins/nativeapp/codegen/snowpark/extension_function.py

```diff
@@ -0,0 +1,343 @@
+import os
+from enum import Enum
+from pathlib import Path
+from textwrap import dedent
+from typing import (
+    Any,
+    Dict,
+    List,
+    Literal,
+    Optional,
+    Tuple,
+    Union,
+)
+
+from snowflake.cli.api.console import cli_console as cc
+from snowflake.cli.plugins.nativeapp.artifacts import NotInDeployRootError
+from snowflake.cli.plugins.nativeapp.codegen.constants import (
+    STAGE_PREFIX,
+    TEMP_OBJECT_NAME_PREFIX,
+)
+from snowflake.cli.plugins.nativeapp.utils import is_single_quoted
+
+
+class TempObjectType(Enum):
+    """Replicates snowflake.snowpark._internal.utils.TempObjectType"""
+
+    TABLE = "TABLE"
+    VIEW = "VIEW"
+    STAGE = "STAGE"
+    FUNCTION = "FUNCTION"
+    FILE_FORMAT = "FILE_FORMAT"
+    QUERY_TAG = "QUERY_TAG"
+    COLUMN = "COLUMN"
+    PROCEDURE = "PROCEDURE"
+    TABLE_FUNCTION = "TABLE_FUNCTION"
+    DYNAMIC_TABLE = "DYNAMIC_TABLE"
+    AGGREGATE_FUNCTION = "AGGREGATE_FUNCTION"
+    CTE = "CTE"
+
+
+def get_handler_path_without_suffix(file_path: Path, deploy_root: Path) -> str:
+    if file_path.is_relative_to(deploy_root):
+        norm_parent_path = os.path.normpath(file_path.parent)
+        parent_paths = norm_parent_path.split(os.path.sep)
+        file_name = file_path.name.split(".")[0]
+        return f"{'.'.join(parent_paths)}.{file_name}"
+    else:
+        raise NotInDeployRootError(
+            artifact_src="_",
+            dest_path=file_path,
+            deploy_root=deploy_root,
+        )
+
+
+def get_object_type_as_text(object_type: TempObjectType) -> str:
+    return object_type.value.replace("_", " ")
+
+
+class ExtensionFunctionProperties:
```

Comment:
The data from snowpark is not enough to create the SQL statements. For example:
1. We need access to the destination file (in deploy root) to create the handler. 
2. We need to validate the raw imports with access to the user's filesystem.
3. `native_app_params` is used to set the value to `self.schema` and `self.application_roles`. When not found, it needs to warn the user accordingly.

and so on.

All this enrichment and validation needs to happen before the SQL can be created. And because they are all related to the snowpark data itself, and have nothing to do with SQL just yet, I wanted to centralize them in one place, i.e. ExtensionFunctionProperties.